detox/src/client/actions/SyncStatusFormatter.js

```diff
@@ -0,0 +1,277 @@
+  }
+
+  static isValidJSONStatus(jsonStatus) {
+    const ajv = new Ajv();
+    const isValidStatus = ajv.compile(SyncStatusFormatter.jsonStatusSchema());
+    return isValidStatus(jsonStatus);
+  }
+
+  static jsonStatusSchema() {
+    return {
+      type: 'object',
+      properties: {
+        app_status: { 'enum': ['busy', 'idle'] },
+        busy_resources: {
+          type: 'array',
+          items: {
+            type: 'object',
+            properties: {
+              name: { 'type': 'string' },
+              description: {
+                type: 'object',
+                additionalProperties: true
+              }
+            },
+            required: ['name'],
+            additionalProperties: false
+          }
+        }
+      },
+      required: ['app_status'],
+      additionalProperties: false
+    };
+  }
+
+  static isAppIdle(jsonStatus) {
+    return jsonStatus.app_status === 'idle' ||
+      (jsonStatus.busy_resources !== undefined && !jsonStatus.busy_resources.length);
+  }
+
+  static resourcesDescriptionsFromJSON(jsonDescriptions) {
+    let descriptions = [];
+    for (const jsonDescription of jsonDescriptions) {
+      const description = SyncStatusFormatter.resourceDescriptionFromJSON(jsonDescription);
+      descriptions.push(description);
+    }
+
+    return descriptions;
+  }
+
+  static resourceDescriptionFromJSON(jsonDescription) {
+    const resourceName = jsonDescription.name;
+    const properties = jsonDescription.description;
+
+    switch (resourceName) {
+      case 'delayed_perform_selector':
+        return SyncStatusFormatter.delayedPerformSelectorResourceDescriptionFromProperties(properties);
+
+      case 'dispatch_queue':
+        return SyncStatusFormatter.dispatchQueueResourceDescriptionFromProperties(properties);
+
+      case 'run_loop':
+        return SyncStatusFormatter.runLoopResourceDescriptionFromProperties(properties);
+
+      case 'one_time_events':
+        return SyncStatusFormatter.oneTimeEventDescriptionFromProperties(properties);
+
+      case 'timers':
+        return SyncStatusFormatter.timersResourceDescriptionFromProperties(properties);
+
+      case 'ui':
+        return SyncStatusFormatter.uiResourceDescriptionFromProperties(properties);
+
+      case 'js_timers':
+        return SyncStatusFormatter.jsTimersResourceDescriptionFromProperties(properties);
+
+      case 'network':
+        return SyncStatusFormatter.networkResourceDescriptionFromProperties(properties);
+
+      case 'looper':
+        return SyncStatusFormatter.looperResourceDescriptionFromProperties(properties);
+
+      case 'io':
+        return SyncStatusFormatter.resourceTitle(`Disk I/O activity.`);
+
+      case 'unknown':
+        return SyncStatusFormatter.unknownResourceDescriptionFromProperties(properties);
+
+      default:
+        throw new DetoxRuntimeError(`Given sync status is invalid, cannot find resource name: \`${resourceName}\``);
```

Comment:
You should not use `DetoxRuntimeError` for things that our users don't bear any responsibility. Use `DetoxInternalError`.

Reply:
OK, got you :+1: